backend/src/game/game.gateway.ts

```diff
@@ -0,0 +1,138 @@
+import {
+  SubscribeMessage,
+  WebSocketGateway,
+  WebSocketServer,
+  WsException,
+} from '@nestjs/websockets';
+import { Server, Socket } from 'socket.io';
+import { Game } from './game';
+import { MoveDto } from '@dtos/game/move.dto';
+import { GameOptionsDto } from '@dtos/game/game-options.dto';
+import { MatchesService } from 'src/matches/matches.service';
+import { ConfigService } from '@nestjs/config';
+import { UsersService } from 'src/users/users.service';
+import { SecureGateway, CheckAuth } from 'src/auth/auth.websocket';
+
+@WebSocketGateway({ cors: true, namespace: 'game' })
+export class GameGateway extends SecureGateway {
+  constructor(
+    protected readonly usersService: UsersService,
+    protected readonly configService: ConfigService,
+    private readonly matchService: MatchesService,
+  ) {
+    super('GameGateway', usersService, configService);
+  }
+
+  @WebSocketServer() server: Server;
+
+  games: Map<number, Game> = new Map(); // array of all active games (game state will only be stored in memory, which I think is fine)
+  queues: Map<string, Array<Socket>> = new Map();
+
+  @SubscribeMessage('game-move')
+  @CheckAuth
+  handleMove(client: Socket, move: MoveDto): string {
+    const game = this.games.get(move.gameId);
+    let player = -1;
+    if (!game) {
+      return 'Error: game not found';
+    }
+    if (client === game.players[0]) {
+      player = 0;
+    } else if (client === game.players[1]) {
+      player = 1;
+    } else {
+      return 'Error: not a player';
+    }
+    if (game) {
+      game.move(player, move.dy);
+    }
+    return 'Success';
+  }
+
+  @SubscribeMessage('game-queue')
+  @CheckAuth
+  async handeQueue(
+    client: Socket,
+    gameOptions: GameOptionsDto,
+  ): Promise<string> {
+    const clientUser = this.getAuthenticatedUser(client);
+    let selectedQueue;
+    const gameOptionsString = JSON.stringify(gameOptions);
```

Comment:
If I understand correctly, it is supposed that `homeId` and `awayId` are never `undefined` at the same time, maybe throw an error if they are.

Reply:
That's mostly correct, but the queue system is also intended to handle the other game modes which will be coming in the future. The idea behid using the `gameOptions` object as the key was to handle multiple game modes without having to hard code separate queues for them.

As far as the `id`s go, the idea is that for a challenge the challenger includes the `homeId` and `awayId` in the `gameOptions` and the challenged accepts by sending the same options (a route/gateway could be created to facilitate this if necessary, but I think it can be handled smoothly by the frontend) . Otherwise, the `id`s are not included. If the user includes one `id` but not the other, then the game would be unlikely to be matched (unless someone else sends the same options). I don't think that it would prevent the queue from working, but it would mean that memory would be wasted keeping track of that.

We'll probably want to add a timeout so the user doesn't wait forever if no match is found, so that could potentially solve the memory problem. However, we can also add a check to reject having only one id defined since there is no reason to allow it.

Reply:
Let us discuss in #250